client/state/themes/upload-theme/reducer.js

```diff
@@ -0,0 +1,80 @@
+/**
+ * External dependencies
+ */
+import { combineReducers } from 'redux';
+import { omit } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { createReducer } from 'state/utils';
+import {
+	THEME_UPLOAD_START,
+	THEME_UPLOAD_SUCCESS,
+	THEME_UPLOAD_FAILURE,
+	THEME_UPLOAD_CLEAR,
+	THEME_UPLOAD_PROGRESS,
+} from 'state/action-types';
+
+export const uploadedTheme = createReducer( {}, {
+	[ THEME_UPLOAD_SUCCESS ]: ( state, { siteId, theme } ) => ( {
+		...state,
+		[ siteId ]: theme,
```

Comment:
Yes, I think you are right. I'm loathe to spend too long on it at this stage, though, because we really need to get this upload stuff done and dusted.

I'm going to timebox a quick attempt at it. If it proves slightly tricky we can change it later on.